.github/workflows/main.yml

```diff
@@ -184,9 +168,63 @@ jobs:
           make lint-js-cached
       - name: Run the links linter
         run: make lint-links
+
+      - name: Check dependency structure
+        run: make check-dependencies
+      - name: Check that JSON schema is up-to-date
+        run: make check-jsonschema
+      # The rest of our code is typechecked in the `build` Makefile target, which
+      # is run above.
+      - name: Run the TypeScript typechecker for contrib
+        run: make typecheck-contrib
+      - name: Run the TypeScript typechecker for scripts
+        run: make typecheck-scripts
+      - name: Run the GitHub Actions linter
+        run: make lint-actions
+      # This step runs at the end, since it is common for it to fail in
+      # dependabot PRs, but we still want all other tests above to run
+      # in those cases.
+      - name: Check for duplicate Node dependencies
+        run: yarn dedupe --check
+
+  native-checks:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+
+      - name: Install OS packages
+        uses: awalsh128/cache-apt-pkgs-action@v1.5.0
+        with:
+          packages: graphviz libgraphviz-dev
+          version: 1.0
+          execute_install_scripts: true
+
+      - name: Install additional dependencies
+        run: |-
+          sudo curl -fsSL https://github.com/hadolint/hadolint/releases/download/v2.12.0/hadolint-Linux-x86_64 -o /bin/hadolint && \
+          sudo chmod +x /bin/hadolint && sudo curl -fsSL https://d2lang.com/install.sh | sh -s -- && \
+          sudo apt-get install -y shellcheck
+      - name: Set up Python
+        uses: actions/setup-python@v5
+        with:
+          python-version: '3.10'
+      - name: Install uv
+        uses: astral-sh/setup-uv@v6
+        with:
+          enable-cache: true
+          cache-dependency-glob: 'images/plbase/python-requirements.txt'
+      - name: Install Python dependencies
+        env:
+          UV_SYSTEM_PYTHON: 1
+        run: uv pip install -r images/plbase/python-requirements.txt
+
+      - name: Set up Node
+        uses: actions/setup-node@v4
+        with:
+          node-version: '20'
+          cache: 'yarn'
```

Comment:
The other job has a `yarn install ...` step after this. Is that not needed here for some reason? Or is this why you added the on-the-fly installation elsewhere?

Reply:
This is why I added the on-the-fly installation. `yarn install` is actually quite "slow" in the context of this PR (30 seconds). Downloading "on-the-fly" is only 5-6 seconds.

Reply:
Gotcha. It feels super gross to clutter our Makefile with that. I'd rather inline the installation steps here since they're specific to the Actions workflow. Then you can leave a single comment explaining why this is done in the first place.

Reply:
`yarn` seems pretty annoying about this. `yarn global add` doesn't work until after `yarn` is run`. `yarn dlx` is temporary, and is deleted after the run. I don't know how to do the install step for one package without installing everything.

Reply:
You might just want to use `npm install -g` here. I don't think we actually rely on any Yarn-specific behavior at runtime.

Reply:
Good idea. [488706b](https://github.com/PrairieLearn/PrairieLearn/pull/11909/commits/488706bf7afde95b143cdd7cd5f60adcc008dc87)